internal/controller/customreplicaset_controller.go

```diff
@@ -74,13 +80,241 @@ func (r *CustomReplicaSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
+	if latestRevision == nil {
+		return fmt.Errorf("invalid controller revision to update CR label with")
+	}
+	if cr.Labels == nil {
+		cr.Labels = map[string]string{}
+	}
+
+	cr.Labels["latestRevisionName"] = latestRevision.Name
+	cr.Labels["latestRevisionNumber"] = strconv.FormatInt(latestRevision.Revision, 10)
+	if err := r.Update(ctx, cr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func hashControllerRevisionData(revisionData runtime.RawExtension) (string, error) {
+	jsonData, err := json.Marshal(revisionData)
+	if err != nil {
+		return "", err
+	}
+
+	// Compute SHA256 hash of PodTemplate
+	hash := sha256.Sum256(jsonData)
+
+	// Convert the hash to hexadecimal string
+	hashString := fmt.Sprintf("%x", hash)
+
+	return hashString, nil
+}
+
+func (r *CustomReplicaSetReconciler) getAllControllerRevisions(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, sorted bool) (*v1.ControllerRevisionList, error) {
+	controllerRevisions := &v1.ControllerRevisionList{}
+	if err := r.List(ctx, controllerRevisions, client.InNamespace(cr.Namespace), client.MatchingLabels{"owner": cr.Name}); err != nil {
+		return nil, err
+	}
+
+	// Sort the ControllerRevisions by Revision
+	if sorted {
+		sort.Slice(controllerRevisions.Items, func(i, j int) bool {
+			return controllerRevisions.Items[i].Revision < controllerRevisions.Items[j].Revision
+		})
+	}
+
+	return controllerRevisions, nil
+}
+
+func (r *CustomReplicaSetReconciler) manageControllerRevisionHistory(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet) error {
+	jsonSpec, err := convertCRSpecToJson(cr.Spec)
+	if err != nil {
+		return err
+	}
+
+	newRevisionData := runtime.RawExtension{Raw: jsonSpec}
+	newRevisionHash, err := hashControllerRevisionData(newRevisionData)
+	if err != nil {
+		return err
+	}
+
+	controllerRevisions, err := r.getAllControllerRevisions(ctx, cr, true)
+	if err != nil {
+		return err
+	}
+
+	_, latestRevisionNumber, err := r.getLatestRevision(ctx, cr, controllerRevisions)
+	if err != nil {
+		return err
+	}
+
+	cachedRevision := &v1.ControllerRevision{}
+	cachedRevisionHash := ""
+	if cachedRevisionName, ok := cr.Labels["latestRevisionName"]; ok {
+		if err := r.Get(ctx, client.ObjectKey{Namespace: cr.Namespace, Name: cachedRevisionName}, cachedRevision); err != nil {
+			fmt.Println("Cached label for revision name is invalid, need to perform search for latest revision")
+		}
+
+		cachedRevisionHash, err = hashControllerRevisionData(cachedRevision.Data)
+		if err != nil {
+			fmt.Println("Failed to generate hash for CR labeled controller revision data")
+			return err
+		}
+	}
+
+	// Search if the revision we are trying to create has been cached in the cr, or already exists
+	// in the revision history, if so update that revision history entry to be the latest,
+	// if not create and append a new revision entry
+	if cachedRevisionHash == newRevisionHash {
+		err := r.updateRevisionToLatest(ctx, cr, cachedRevision, latestRevisionNumber)
```

Comment:
If the hash matches, you should do nothing, right?

Reply:
Yeah updateRevisionToLatest, just makes sure the cached version is the latest. It usually won't be doing anything. But in the case say someone manually creates a new controllerrevision, the cached version wouldn't be updated, so we ensure its latest.

Reply:
On second thought though this would be highly unsual...users arent generally doing crud on controller revisions, I fixed this

Reply:
You don't have to ensure latest for controllerrevision because what presented in cr's template is the latest. If someone manually created controllerrevision, it should be ignored because the source of controllerrevision is the cr. Will manually created controllerrevision mess up the controller? Maybe, but we see that as users shot their own feet on purpose,

Reply:
yes you're right, since we added the owner label. Someone would have to not only create a controllerrevision, but create it with that specific label, we're filtering on as well. thanks